File: Web/RMA-Web-Shell/projects/shared-components-lib/src/lib/data-exchange/data-exchange/core/factory/data-exchange-factory.ts
import { dataExchangeModelBase } from "projects/shared-models-lib/src/lib/data-exchange/data-exchange-model";
import { DataExchangeBase } from "../data-exchange";
import { excelFileExchange } from "../fileExchange/excel/excel-file-exchange";
import { excelDataExchangeModel } from "projects/shared-models-lib/src/lib/data-exchange/fileExchange/excel/excel-data-exchange";
import { csvFileExchange } from "../fileExchange/csv/csv-file-exchange";
import { csvDataExchangeModel } from "projects/shared-models-lib/src/lib/data-exchange/fileExchange/csv/csv-data-exchange";

export class DataExchangeFactory {
    CreateDataExchange(dataExchangeModelBase: dataExchangeModelBase): DataExchangeBase {
        if(dataExchangeModelBase instanceof(excelDataExchangeModel))
            return new excelFileExchange();
        if(dataExchangeModelBase instanceof(csvDataExchangeModel))
            return new csvFileExchange();
    }
}

File: Service/Billing/RMA.Service.Billing/DataExchange/Factory/DataExchangeFactory.cs
using RMA.Common.Extensions;
using RMA.Service.Billing.Contracts.Entities.DataExchange;
using RMA.Service.Billing.Contracts.Entities.DataExchange.Import;
using RMA.Service.Billing.Contracts.Entities.FileExchange;
using RMA.Service.Billing.DataExchange.Data;

using System;

namespace RMA.Service.Billing.DataExchange.Factory
{
    /// <summary>
    /// Design pattern: Generic factory pattern
    /// Honor the open-closed principle with the factory and OOP
    /// </summary>
    /// <typeparam name="TExchangeModel"></typeparam>
    public class DataExchangeFactory<TExchangeModel>
        where TExchangeModel : DataExchangeModelBase
    {
        public DataExchangeBase CreateDataExchange()
        {
            if (typeof(TExchangeModel).IsAssignableTo(typeof(FileExchangeModelBase)))
            {
                //Return Excel or CSV etc...
                if (typeof(TExchangeModel) == typeof(BillingDataExchangeModel))
                    return new BillingDataExchange();
            }

            throw new InvalidOperationException();
        }
    }
}